pyro/distributions/util.py

```diff
@@ -98,6 +98,31 @@ def sum_rightmost(value, dim):
     return value.contiguous().view(value.shape[:-dim] + (-1,)).sum(-1)
 
 
+def sum_leftmost(x, dim):
+    """
+    Sum all but a certain number of rightmost dimensions of a given tensor ``x``.
+
+    If ``dim`` is 0, no dimensions are summed out.
+    If ``dim`` is -1, all but the rightmost dimension is summed out.
+    If ``dim`` is -2, all but the two rightmost dimensions are summed out.
+    etc.
+
+    Example
+    ```
+    x = torch.ones(2,3,4)
+    assert sum_leftmost(x, -2).shape == (3, 4)
+    assert sum_leftmost(x, -1).shape == (4,)
+    ```
+
+    :param torch.autograd.Variable x: A tensor
+    :param int dim: Specifies the number of dims to sum out (a non-positive integer)
+    """
+    assert(dim <= 0), "dim must be less than or equal to zero"
```

Comment:
Could you please implement the general interface I suggested.
```py
def sum_leftmost(x, dim):
    if dim < 0:
        dim += x.dim()
    ...
```
and add a test
```py
def test_sum_leftmost():
    x = torch.ones(2,3,4)
    assert sum_leftmost(x, 0).shape == (2, 3, 4)
    assert sum_leftmost(x, 1).shape == (3, 4)
    assert sum_leftmost(x, 2).shape == (4,)
    assert sum_leftmost(x, -1).shape == (4,)
    assert sum_leftmost(x, -2).shape == (3, 4)
```

Reply:
i didn't understand your interface

Reply:
since ``dim += x.dim()`` was different. i guess now i see it

Reply:
Sorry for being unclear. The interface should match `sum_rightmost()`: for positive integers `n`, `sum_leftmost(x, n)` sums out the `n` leftmost dimensions. For negative integers `-n`, `sum_leftmost(x, -n)` sums out all but the `n` rightmost dimensions, i.e. the leftmost `x.dim() - n` dimensions. This matches the behavior of negative indexing for many other operations.

Reply:
Thanks, I'm only requesting this because we'll be needing it in our work on `iarange` and parallel enumeration, so your efforts in this PR will pay off.